File: exercicios/projeto-final/skeletons/pokeapi/flutter/pratica/lib/api/poke_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/pokemon.dart';

class PokeApi {
  static const String baseUrl = 'https://pokeapi.co/api/v2';

  Future<List<PokemonSummary>> fetchList({int limit = 151}) async {
    final res =
        await http.get(Uri.parse('$baseUrl/pokemon?limit=$limit&offset=0'));
    if (res.statusCode != 200) {
      throw Exception('Falha ao carregar lista de pokémons');
    }
    final body = jsonDecode(res.body) as Map<String, dynamic>;
    final results = body['results'] as List;
    return results
        .map((e) => PokemonSummary.fromListEntry(
              e['name'] as String,
              e['url'] as String,
            ))
        .toList();
  }

  Future<PokemonDetail> fetchDetail(int id) async {
    final res = await http.get(Uri.parse('$baseUrl/pokemon/$id'));
    if (res.statusCode != 200) {
      throw Exception('Falha ao carregar detalhe do pokémon $id');
    }
    return PokemonDetail.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  /// Retorna o conjunto de nomes de pokémon que pertencem ao tipo [type].
  Future<Set<String>> fetchNamesByType(String type) async {
    final res = await http.get(Uri.parse('$baseUrl/type/$type'));
    if (res.statusCode != 200) {
      throw Exception('Falha ao carregar tipo $type');
    }
    final body = jsonDecode(res.body) as Map<String, dynamic>;
    final entries = body['pokemon'] as List;
    return entries
        .map((e) => e['pokemon']['name'] as String)
        .toSet();
  }
}

File: exercicios/projeto-final/skeletons/pokeapi/react-native/pratica/src/api/pokeApi.ts
import { PokemonDetail, PokemonSummary, idFromUrl } from '../types/pokemon';

const BASE_URL = 'https://pokeapi.co/api/v2';

export async function fetchList(limit = 151): Promise<PokemonSummary[]> {
  const res = await fetch(`${BASE_URL}/pokemon?limit=${limit}&offset=0`);
  if (!res.ok) throw new Error('Falha ao carregar lista de pokémons');
  const body = await res.json();
  return body.results.map((entry: { name: string; url: string }) => ({
    id: idFromUrl(entry.url),
    name: entry.name,
  }));
}

export async function fetchDetail(id: number): Promise<PokemonDetail> {
  const res = await fetch(`${BASE_URL}/pokemon/${id}`);
  if (!res.ok) throw new Error(`Falha ao carregar detalhe do pokémon ${id}`);
  const body = await res.json();
  return {
    id: body.id,
    name: body.name,
    types: body.types.map((t: { type: { name: string } }) => t.type.name),
    heightDm: body.height,
    weightHg: body.weight,
  };
}

export async function fetchNamesByType(type: string): Promise<Set<string>> {
  const res = await fetch(`${BASE_URL}/type/${type}`);
  if (!res.ok) throw new Error(`Falha ao carregar tipo ${type}`);
  const body = await res.json();
  return new Set(
    body.pokemon.map((entry: { pokemon: { name: string } }) => entry.pokemon.name),
  );
}
